include/ginkgo/core/base/lin_op.hpp

```diff
@@ -767,6 +767,58 @@ class EnableAbsoluteComputation : public AbsoluteComputable {
 };
 
 
+/**
+ * Adds the operation M <- a I + b M for matrix M, identity
+ * operator I and scalars a and b, where M is the calling object.
+ */
+class ScaledIdentityAddable {
+public:
+    /**
+     * Scales this and adds another scalar times the identity to it.
+     *
+     * @param a  Scalar to multiply the identity operator by before adding.
+     * @param b  Scalar to multiply this before adding the scaled identity to
+     *   it.
+     */
+    virtual void add_scaled_identity(const LinOp* const a,
+                                     const LinOp* const b) = 0;
+};
+
+
+/**
+ * Implements common checks for the operation M <- a I + b M for matrix M,
+ * identity operator I and scalars a and b, where M is the calling object.
+ *
+ * @tparam ConcreteType  The concrete type that should have this operation
+ *                       (the CRTP parameter).
+ */
+template <typename ConcreteType>
+class EnableScaledIdentityAddition : public ScaledIdentityAddable {
+public:
+    /**
+     * Scales this and adds another scalar times the identity to it.
+     *
+     * @param a  Scalar to multiply the identity operator by before adding.
+     * @param b  Scalar to multiply this before adding the scaled identity to
+     *   it.
+     */
+    void add_scaled_identity(const LinOp* const a,
+                             const LinOp* const b) override
+    {
+        GKO_ASSERT_IS_SCALAR(a);
+        GKO_ASSERT_IS_SCALAR(b);
+        auto ae =
+            make_temporary_clone(as<ConcreteType>(this)->get_executor(), a);
```

Comment:
nit: usually we use a `self` function that `static_cast`s to ConcreteType*

Reply:
Perhaps it's best to just replace `as` with `static_cast` instead of adding two more identifiers in the interface?

Reply:
True, if we only use it in a single place, we don't need to add another function.